rllab/tf/algos/ddpg.py

```diff
@@ -0,0 +1,319 @@
+            target_actor=target_actor)
+
+    def _learn(self):
+        """
+        Perform algorithm optimizing.
+
+        Returns:
+            action_loss: Loss of action predicted by the policy network.
+            qval_loss: Loss of q value predicted by the q network.
+            ys: y_s.
+            qval: Q value predicted by the q network.
+        """
+        replay_buffer = self.opt_info["replay_buffer"]
+        target_actor = self.opt_info["target_actor"]
+        target_critic = self.opt_info["target_critic"]
+        f_train_critic = self.opt_info["f_train_critic"]
+        f_train_actor = self.opt_info["f_train_actor"]
+
+        transitions = replay_buffer.random_sample(self.batch_size)
+        observations = transitions["observations"]
+        rewards = transitions["rewards"]
+        actions = transitions["actions"]
+        terminals = transitions["terminals"]
+        next_observations = transitions["next_observations"]
+        rewards = rewards.reshape(-1, 1)
+        terminals = terminals.reshape(-1, 1)
+
+        next_actions = target_actor.get_actions(next_observations)
+        next_qvals = target_critic.get_qval(next_observations, next_actions)
+
+        ys = rewards + (1. - terminals) * self.discount * next_qvals
+
+        _, action_loss = f_train_actor(observations)
+        _, qval_loss, qval = f_train_critic(ys, observations, actions)
+
+        return action_loss, qval_loss, ys, qval
+
+    def _predict(self, state):
+        """
+        Select action a_t = mu(s_t) + N according to the input state and
+        exploration strategy.
+
+        Args:
+            state: Current env observation.
+
+        Returns:
+            action: Sampled action.
+        """
+        action = self.actor.get_action(np.array(state))
+        noise = self.es.get()
+        action = action + noise
+        action = np.clip(action, self.action_range[0], self.action_range[1])
+        return action
+
+    @overrides
+    def train(self, sess=None):
+        self._initialize()
+
+        replay_buffer = self.opt_info["replay_buffer"]
+        created_session = True if (sess is None) else False
+        if sess is None:
+            sess = tf.Session()
+            sess.__enter__()
+
+        sess.run(tf.global_variables_initializer())
+
+        state = self.env.reset()
+        if self.es:
+            self.es.reset()
+
+        episode_rewards_history = deque(maxlen=100)
+        episode_reward = 0.
+        episode_step = 0
+        episodes = 0
+        epoch_episode_rewards = []
+        epoch_episode_steps = []
+        epoch_actions = []
+        epoch_actor_losses = []
+        epoch_critic_losses = []
+        epoch_ys = []
+        epoch_qs = []
+        epoch_episodes = 0
+
+        for epoch in range(self.n_epochs):
+            for step in range(self.n_epoch_cycles):
+                for rollout in range(self.n_rollout_steps):
+                    if self.plot:
+                        self.env.render()
```

Comment:
Please use the tf async plotter if possible because synchronized plotting will slow down the training process.